includes/class-wcsr-resource.php

```diff
@@ -187,24 +188,84 @@ public function get_days_active( $from_timestamp, $to_timestamp = null ) {
+			}
+
+			array_unshift( $activation_times, $start_timestamp );
 		}
 
 		foreach ( $activation_times as $i => $activation_time ) {
-			if ( isset( $deactivation_times[ $i ] ) ) {
-				$days_active -= floor( ( $activation_time - $deactivation_times[ $i ] ) / DAY_IN_SECONDS );
+			// If there is corresponding deactivation timestamp, the resource has deactivated before the end of the period so that's the time we want, otherwise, use the end of the period as the resource was still active at end of the period
+			$deactivation_time = isset( $deactivation_times[ $i ] ) ? $deactivation_times[ $i ] : $to_timestamp;
+
+			// skip over any days that are activated/deactivated on the same day and have already been accounted for
+			if ( $i !== 0 && self::is_on_same_day( $deactivation_time, $deactivation_times[ $i - 1 ], $from_timestamp ) ) {
+				continue;
+			}
+
+			// Calculate days based on time between
+			$days_by_time = intval( ceil( ( $deactivation_time - $activation_time ) / DAY_IN_SECONDS ) );
+
+			// Increase our tally
+			$days_active += $days_by_time;
+
+			// If days based on time is only 1 but it was "across a day" we may need to adjust IF NOT accounted for already
+			if ( $days_by_time == 1 && ! self::is_on_same_day( $activation_time, $deactivation_time, $from_timestamp ) ) {
+
+				// handle situation if first activation crosses a day
+				if ( $i == 0 && ! self::is_on_same_day( $activation_time, $deactivation_time, $from_timestamp ) ) {
+					$days_active += 1;
+				}
+
+				// if this activation didn't start on the same day as previous activation it is safe to add an extra day
+				if ( $i !== 0 && ! self::is_on_same_day( $activation_time, $deactivation_times[ $i - 1 ], $from_timestamp ) ) {
+					$days_active += 1;
+				}
 			}
 		}
 
 		return $days_active;
 	}
 
 	/**
+	 * Prototype/demo conditional check for whether a timestamp is on the same "day" as another timestamp
+	 *
+	 * The catch is the "day" is not typical calendar day - it based on a 24 hour period from the $start_timestamp
+	 *
+	 * The $start_timestamp could be the actual time it was first activated of the start of the period.
+	 *
+	 * Takes the starting timestamp and gets the time our days start from.
+	 * Takes the timestamp we are checking if we are on the same day as and gets the date, previous days date and time
+	 * Works out whether the current day started on the same date as the one we are comparing with or a day earlier and works makes a time stampe fo when the day starts
+	 * Works out when the days ends
+	 * Check if our timestamp is with the start and end dates we have determined
+	 *
+	 * @param  int  $current_timestamp    [description]
+	 * @param  int  $compare_timestamp [description]
+	 * @param  int  $start_timestamp  [description]
+	 * @return boolean true on same day | false if not
+	 */
+	public static function is_on_same_day( $current_timestamp, $compare_timestamp, $start_timestamp ) {
```

Comment:
Does this need to be public?

It shouldn't be public just for unit testing (and it can still be tested even if set to `protected`, see [this for an example](https://github.com/Prospress/woocommerce-subscriptions/blob/6d659b2044c5041f3e339b4beb7bc9f3bba09654/tests/framework/class-wcs-unit-test-case.php#L173:L188) of how to do that).

Reply:
I dont see why it needs to be public. I'll update this and tests as well :+1: